rust/scx_utils/src/autopower.rs

```diff
@@ -14,18 +18,56 @@ trait PowerProfiles {
     fn active_profile(&self) -> Result<String>;
 }
 
-static POWER_PROFILES_PROXY: OnceLock<Option<PowerProfilesProxyBlocking<'static>>> =
-    OnceLock::new();
+static POWER_PROFILES_PROXY: OnceLock<PowerProfilesProxyBlocking<'static>> = OnceLock::new();
+static RETRIES: AtomicUsize = AtomicUsize::new(0);
+const MAX_RETRIES: usize = 10;
+
+fn read_energy_profile() -> String {
+    let energy_pref_path = "/sys/devices/system/cpu/cpufreq/policy0/energy_performance_preference";
+    let scaling_governor_path = "/sys/devices/system/cpu/cpufreq/policy0/scaling_governor";
 
-pub fn fetch_power_profile() -> String {
-    POWER_PROFILES_PROXY
-        .get_or_init(|| {
-            let system_bus = Connection::system().ok()?;
-            let system_bus = Box::leak(Box::new(system_bus));
-            PowerProfilesProxyBlocking::new(system_bus).ok()
+    fs::read_to_string(energy_pref_path)
+        .ok()
+        .or_else(|| fs::read_to_string(scaling_governor_path).ok())
+        .and_then(|s| {
+            Some(match s.trim_end() {
+                "power" | "balance_power" | "powersave" => "power-saver",
+                "balance_performance" => "balanced",
+                "performance" => "performance",
+                _ => return None,
+            })
         })
-        .as_ref()
-        .map(|power_profiles_proxy| power_profiles_proxy.active_profile().ok())
-        .flatten()
         .unwrap_or_default()
+        .to_string()
+}
+
+pub fn fetch_power_profile(no_ppd: bool) -> String {
+    if no_ppd || std::env::var("SCX_NO_PPD").is_ok_and(|s| s == "1") {
```

Comment:
Sorry for nitpicking, but with the `no_ppd` option isn't the `SCX_NO_PPD` env variable redundant? Does it bring any benefit having an env variable to override the option?

That aside, LGTM.